### PR TITLE
Support the Bestiary + new NPC Loot system

### PR DESCRIPTION
**Pull request by kittenchilly**
### What is the new feature?
This PR adds support to the new Bestiary introduced in 1.4. It automatically adds any modded NPCs into the Bestiary list (unless specified otherwise in SetStaticDefaults), as well as adding a new SetBestiary hook in ModNPC & GlobalNPC that allows you to set Bestiary tags, spawning conditions, and descriptions.
Alongside that, this PR adds support for the extensive new NPC Loot system with a new ModifyNPCLoot hook (the old NPCLoot hooks were renamed to OnDeath)..

### Why should this be part of tModLoader?
Currently opening the Bestiary crashes in the current build, technically making this a bug fix. I did both the Bestiary and the new NPC Loot system in one PR as they are closely linked to eachother functionally.

### ExampleMod updates
Re-added the ExampleWorm NPC, mostly as an example of Bestiary formatting for any multi-part NPCs.
Updated both the PartyZombie and ExamplePerson NPCs as well to have new Bestiary and NPC Loot info.
Added extensive GlobalNPC documentation for practically everything you can do with the new NPC Loot system.

### CODE REVIEW

File: patches/tModLoader/Terraria/GameContent/Bestiary/BestiaryDatabaseNPCsPopulator.cs.patch
Comment (Mirsario) [inline patches/tModLoader/Terraria/GameContent/Bestiary/BestiaryDatabaseNPCsPopulator.cs.patch:40]: Unnecessarily big patch. Just use a return.

**Review by Mirsario — CHANGES_REQUESTED** [review]

File: patches/tModLoader/Terraria/GameContent/Bestiary/BestiaryDatabaseNPCsPopulator.cs.patch
Comment (Jofairden) [inline patches/tModLoader/Terraria/GameContent/Bestiary/BestiaryDatabaseNPCsPopulator.cs.patch:109]: Why was the method renamed?

File: patches/tModLoader/Terraria/GameContent/UI/Elements/UIBestiaryEntryGrid.cs.patch
Comment (Jofairden) [inline patches/tModLoader/Terraria/GameContent/UI/Elements/UIBestiaryEntryGrid.cs.patch:8]: Curious as to why this has to change

File: patches/tModLoader/Terraria/GameContent/Bestiary/BestiaryDatabaseNPCsPopulator.cs.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/Bestiary/BestiaryDatabaseNPCsPopulator.cs.patch:124]: Looks good, just see Jof and Mirsario's comments above.

File: patches/tModLoader/Terraria/GameContent/Bestiary/UnlockableNPCEntryIcon.cs.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/Bestiary/UnlockableNPCEntryIcon.cs.patch:16]: please put a //TODO on this, to re-evaluate when we get the asset system working better

File: patches/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropDatabase.cs.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropDatabase.cs.patch:16]: reason for access change? I don't see this used by any examples

File: patches/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropRule.cs.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropRule.cs.patch:8]: Need an example using the `dropsXoutofY` parameter, and a description of how it works.

Comment (Chicken-Bones) [reply patches/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropRule.cs.patch:8]: Also not sure if vanilla has flubbed here, but the parameter to `Common` `dropsOutOfX` is called `dropsOutOfY` in `CommonDrop` looks like vanilla wants nice integer probabilities like 1/Y, and you're exposing the extra parameter for things like 2/7

File: patches/tModLoader/Terraria/GameContent/UI/Elements/UIBestiaryEntryGrid.cs.patch
Comment (Chicken-Bones) [reply patches/tModLoader/Terraria/GameContent/UI/Elements/UIBestiaryEntryGrid.cs.patch:8]: Redundant patch, please remove. Looks like you previously had something at the bottom of this method but since reworked it

File: patches/tModLoader/Terraria/GameContent/UI/Elements/UIBestiaryEntryGrid.cs.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/UI/Elements/UIBestiaryEntryGrid.cs.patch:17]: Please leave a comment on why this changed for future maintainers, since it's not obviously related to other changes in this class

File: patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/UI/EmoteBubble.cs.patch:14]: This won't work, because `NPCCount` will change when mods are loaded and unloaded. You need to reference this in `NPCLoader.ResizeArrays`. Keep this in mind for any arrays active when mods are loaded and do the same for ItemLoader and etc. The only time this will work is if it's a non-static field and the array is constructed and destroyed during gameplay. (that is, when the array lifetime doesn't overlap with mod loading)

File: patches/tModLoader/Terraria/ID/NPCID.cs.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ID/NPCID.cs.patch:25]: This is far from ideal, please find another solution, even if it means some hardcode string manipulation where `CustomTexturePath` is used, and a `TODO` message

File: patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs:205]: All of this looks good

**Review by Chicken-Bones — COMMENTED** [review]
Only a couple of changes. Well done. Definitely need more content for the loot system in example mod, but that can be in future PRs

File: patches/tModLoader/Terraria/GameContent/Bestiary/UnlockableNPCEntryIcon.cs.patch
Comment (kittenchilly) [reply patches/tModLoader/Terraria/GameContent/Bestiary/UnlockableNPCEntryIcon.cs.patch:16]: Yeah, that was the plan.

File: patches/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropRule.cs.patch
Comment (kittenchilly) [reply patches/tModLoader/Terraria/GameContent/ItemDropRules/ItemDropRule.cs.patch:8]: Its kinda odd. Vanilla is contradictory with how it uses this. I definitely need to do more testing with it once I do ExampleMod stuff.

File: patches/tModLoader/Terraria/ID/NPCID.cs.patch
Comment (kittenchilly) [reply patches/tModLoader/Terraria/ID/NPCID.cs.patch:25]: This is the only way that I could get it to actually work with the current state of the asset system. If I just did a string.Replace thing turning Images/ into Terraria/, then it would also affect mods where im certain some will have an images folder they put everything in, which doesnt really make it feasible. I do not plan for this to final. I'll definitely add a comment.

Comment (Chicken-Bones) [reply patches/tModLoader/Terraria/ID/NPCID.cs.patch:25]: `if (CustomTexturePath.StartsWith("Images/")) CustomTexturePath = "Terraria" + CustomTexturePath.SubString(6);`

works unless a mod is named images, since mod name has to come first

File: ExampleMod/Content/NPCs/PartyZombie.cs
Comment (Jofairden) [reply ExampleMod/Content/NPCs/PartyZombie.cs:21]: Man they used such odd wording here, I'd love just an `NPCID.Sets.NPCBestiaryDrawModifiers.AnimateX[id] = 1f` or something

File: ExampleMod/Content/NPCs/PartyZombie.cs
Comment (Mirsario) [inline ExampleMod/Content/NPCs/PartyZombie.cs:45]:
```suggestion
			database.RegisterToNPC(npc.type, ItemDropRule.Common(ItemID.Shackle, 50)); //Drop shackles with a 1 out of 50 chance.
			database.RegisterToNPC(npc.type, ItemDropRule.Common(ItemID.ZombieArm, 250)); //Drop zombie arm with a 1 out of 250 chance.	
```

File: ExampleMod/Content/NPCs/PartyZombie.cs
Comment (Mirsario) [inline ExampleMod/Content/NPCs/PartyZombie.cs:58]: We need this to be localizeable somehow

File: ExampleMod/Content/NPCs/Worm.cs
Comment (Mirsario) [inline ExampleMod/Content/NPCs/Worm.cs:25]:
```suggestion
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Example Worm");
	
			var drawModifier = new NPCID.Sets.NPCBestiaryDrawModifiers(0) { //Influences how the NPC looks in the Bestiary
				CustomTexturePath = "ExampleMod/Content/NPCs/ExampleWorm_Bestiary", //If the NPC is multiple parts like a worm, a custom texture for the Bestiary is encouraged.
				Position = new Vector2(40f, 24f),
				PortraitPositionXOverride = 0f,
				PortraitPositionYOverride = 12f
			};
			NPCID.Sets.NPCBestiaryDrawOffset.Add(npc.type, drawModifier);
```

File: ExampleMod/Content/NPCs/Worm.cs
Comment (Mirsario) [inline ExampleMod/Content/NPCs/Worm.cs:126]:
```suggestion
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Example Worm");
		}
```

File: patches/tModLoader/Terraria/ID/NPCID.cs.patch
Comment (Mirsario) [inline patches/tModLoader/Terraria/ID/NPCID.cs.patch:35]: There have to be smarter solutions.

File: patches/tModLoader/Terraria/ModLoader/ModContent.cs
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/ModContent.cs:457]:
```suggestion
		private static void SetupBestiary(CancellationToken token) {
			//Beastiary DB
			var bestiaryDatabase = new BestiaryDatabase();
			new BestiaryDatabaseNPCsPopulator().Populate(bestiaryDatabase);
			Main.BestiaryDB = bestiaryDatabase;
			ContentSamples.RebuildBestiarySortingIDsByBestiaryDatabaseContents(bestiaryDatabase);
			
			//Drops DB
			var itemDropDatabase = new ItemDropDatabase();
			itemDropDatabase.Populate();
			Main.ItemDropsDB = itemDropDatabase;
			
			//Update the bestiary DB with the drops DB.
			bestiaryDatabase.Merge(Main.ItemDropsDB);
			
			//Etc
			Main.BestiaryUI = new UIBestiaryTest(Main.BestiaryDB);
			Main.ItemDropSolver = new ItemDropResolver(itemDropDatabase);
			Main.BestiaryTracker = new BestiaryUnlocksTracker();
		}
```

File: ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs
Comment (Mirsario) [inline ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs:19]: I would like to see the reason why we can't have registration look something like the following, be set in stone in this PR's thread.

```cs
npcDrops.Add(ItemDropRule.Common(1, 1));
```

**Review by Mirsario — CHANGES_REQUESTED** [review]
The texture path patches need to be looked into, but this is good overall!

File: ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs
Comment (kittenchilly) [reply ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs:19]: Most modded NPC Loot stuff using this method is handled by vanilla with no actual repercussions of doing so (like saving, etc), so making registration different and use a specific method to tmodloader would be difficult, unnecessary, and put modded loot at risk of not doing something vanilla does.

File: patches/tModLoader/Terraria/ID/NPCID.cs.patch
Comment (kittenchilly) [reply patches/tModLoader/Terraria/ID/NPCID.cs.patch:35]: Easiest way without causing file path issues for mods. This was never meant to be permanent.

File: ExampleMod/Content/NPCs/PartyZombie.cs
Comment (kittenchilly) [reply ExampleMod/Content/NPCs/PartyZombie.cs:58]: It already is.

File: ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs
Comment (Mirsario) [reply ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs:19]: Not necessary. I imagined just providing a readonly structure that would have private NPC id and ItemDropDatabase fields, with redirect methods on it.

Comment (Mirsario) [reply ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs]: Implemented what I was talking about.

File: patches/tModLoader/Terraria/ID/NPCID.cs.patch
Comment (Mirsario) [reply patches/tModLoader/Terraria/ID/NPCID.cs.patch]: Reduced the amount of patches to 1, and also removed the need to remove 'Images' prefix.

**Review by Mirsario — APPROVED** [review]

File: ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs
Comment (Chicken-Bones) [inline ExampleMod/Common/GlobalNPCs/ExampleNPCLoot.cs:27]: Fix this todo before merge

File: ExampleMod/Content/NPCs/ExamplePerson.cs
Comment (Chicken-Bones) [inline ExampleMod/Content/NPCs/ExamplePerson.cs:67]: remove the 3 here

File: ExampleMod/Content/NPCs/PartyZombie.cs
Comment (Chicken-Bones) [inline ExampleMod/Content/NPCs/PartyZombie.cs:53]: either use 2 Add calls or one AddRange

File: ExampleMod/Content/NPCs/Worm.cs
Comment (Chicken-Bones) [inline ExampleMod/Content/NPCs/Worm.cs:37]: as above, remove 2. You can probably even do `new []`

File: patches/tModLoader/Terraria/GameContent/Creative/ItemsSacrificedUnlocksTracker.cs.patch
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/GameContent/Creative/ItemsSacrificedUnlocksTracker.cs.patch:15]: Why?

File: patches/tModLoader/Terraria/ModLoader/NPCLoader.cs
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/NPCLoader.cs:118]: foreach on players?

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]
Only minor syntax/access questions, and an example of `NPCLoot.RemoveWhere`

**Pull request merged**
